flink-runtime/src/main/java/org/apache/flink/runtime/scheduler/declarative/DeclarativeScheduler.java

```diff
@@ -471,44 +478,91 @@ private ParallelismAndResourceAssignments determineParallelismAndAssignResources
 
         final Collection<SlotInfoWithUtilization> freeSlots =
                 declarativeSlotPool.getFreeSlotsInformation();
-        final int slotsPerVertex = freeSlots.size() / jobGraph.getNumberOfVertices();
+        // TODO: This can waste slots if the max parallelism for slot sharing groups is not equal
+        final int slotsPerSlotSharingGroup =
+                freeSlots.size() / jobGraph.getSlotSharingGroups().size();
         final Iterator<SlotInfoWithUtilization> slotIterator = freeSlots.iterator();
 
-        for (JobVertex jobVertex : jobGraph.getVerticesSortedTopologicallyFromSources()) {
-            final int parallelism = Math.min(jobVertex.getParallelism(), slotsPerVertex);
-            jobVertex.setParallelism(parallelism);
+        for (SlotSharingGroup slotSharingGroup : jobGraph.getSlotSharingGroups()) {
+            final List<JobVertex> containedJobVertices =
+                    slotSharingGroup.getJobVertexIds().stream()
+                            .map(jobGraph::findVertexByID)
+                            .collect(Collectors.toList());
 
-            parallelismPerJobVertex.put(jobVertex.getID(), parallelism);
+            final Map<Integer, Set<ExecutionVertexID>> sharedSlotToVertexAssignment =
+                    determineParallelismAndAssignToFutureSlotIndex(
+                            containedJobVertices,
+                            slotsPerSlotSharingGroup,
+                            parallelismPerJobVertex);
```

Comment:
I think we also need something similar for when we declare the required resources.